Tallies/TallyClerks/MGxsClerk_class.f90

```diff
@@ -0,0 +1,868 @@
+    k      = 1    ! to calculate transport xss
+    delta  = ZERO
+    deltaStd = ZERO
+
+    ! Loop over all energies and spatial bins
+    do i = 1, N*M
+
+      ! Retrieve reaction rates and flux from memory
+      addr = self % getMemAddress() + self % width * (i - 1) - 1
+      call mem % getResult(flux,  fluxStd,  addr + FLUX_idx)
+      call mem % getResult(fiss,  fissStd,  addr + FISS_idx)
+      call mem % getResult(capt,  captStd,  addr + CAPT_idx)
+      call mem % getResult(scatt, scattStd, addr + SCATT_idx)
+      call mem % getResult(nu,    nuStd,    addr + NUFISS_idx)
+      call mem % getResult(chi,   chiStd,   addr + CHI_idx)
+      call mem % getResult(scattProb, scattProbStd, addr + SCATT_EV_idx)
+
+      ! Calculate MG constants, being careful to avoid division by zero
+      ! If flux is zero all cross sections must be set to zero
+      if (flux == ZERO) then
+        capt_res(1:2,i) = ZERO
+        scattXS         = ZERO
+        scattXSstd      = ZERO
+      else
+        capt_res(1,i) = capt/flux
+        capt_res(2,i) = capt_res(1,i) * sqrt((captStd/capt)**2 + (fluxStd/flux)**2)
+        scattXS       = scatt/flux
+        scattXSstd    = scattXS * sqrt((scattStd/scatt)**2 + (fluxStd/flux)**2)
+      end if
+
+      ! Calculate fission production term and uncertainties
+      if (fiss == ZERO) then
+        fiss_res(1:2,i) = ZERO
+        nu_res(1:2,i)   = ZERO
+      else
+        fiss_res(1,i) = fiss/flux
+        fiss_res(2,i) = fiss_res(1,i) * sqrt((fissStd/fiss)**2 + (fluxStd/flux)**2)
+        nu_res(1,i)   = nu/fiss
+        nu_res(2,i)   = nu_res(1,i) * sqrt((nuStd/nu)**2 + (fissStd/fiss)**2)
+      end if
+
+      ! Store fission spectrum
+      chi_res(1,i) = chi
+      chi_res(2,i) = chiStd
+      sumChi = sumChi + chi
+      ! If this is the last energy group for a material, normalise the spectrum
+      if (mod(i,N) == 0) then
+        g1   = i+1-N
+        gEnd = i
+        if (sumChi /= ZERO) chi_res(1:2, g1:gEnd) = chi_res(1:2, g1:gEnd)/sumChi
+        sumChi = 0
+      end if
+
+      ! Store total cross section and flux for this energy group
+      tot(i)    = capt_res(1,i) + fiss_res(1,i) + scattXS
+      totStd(i) = sqrt(capt_res(2,i)**2 + fiss_res(1,i)**2 + scattXSstd**2)
+      fluxG(i)  = flux
+      fluxGstd(i) = fluxStd
+
+      ! Loop over outgoing energies to calculate the scattering matrices
+      do j = 1, N
+
+        ! Retrieve results from memory
+        call mem % getResult(P0, P0std, addr + SCATT_EV_idx + j)
+        call mem % getResult(P1, P1std, addr + SCATT_EV_idx + N + j)
+        call mem % getResult(prod, prodStd, addr + SCATT_EV_idx + 2*N + j)
+
+        ! Calculate scattering matrices and uncertainties
+        idx = N*(i-1) + j
+        if (P0 == ZERO) then
+          P0_res(1:2, idx) = ZERO
+          P1_res(1:2, idx) = ZERO
+          prod_res(1, idx) = ONE
+          prod_res(2, idx) = ZERO
+        else
+          P0_res(1, idx) = P0/scattProb*scattXS
```

Comment:
I probably don't understand this well enough but... Shouldn't P0 just be 1? I say because it feels like neutron weight is being counted multiple times with P0 and scattXS which doesn't feel intuitive to me.